### PR TITLE
(HTCONDOR-2163)  New Python bindings, initial merge follow-up.

### PR DESCRIPTION
**Pull request by Todd-L-Miller**
A number of items, some large, some small, were punted to HTCONDOR-2163 to avoid unnecessarily delaying the merge of HTCONDOR-1598.

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [ ] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: bindings/python/htcondor2/_collector.py [hunk 1]
```diff
@@ -28,12 +28,19 @@ def _ad_type_from_daemon_type(daemon_type: DaemonType):
...
 
     #
     # In version 1, there was a distinct DaemonLocation type (a named tuple)
     # that `pool` could also be, but that functionality was never documented.
     #
     def __init__(self, pool = None):
+        """
+
+        :param pool:
+        """
```
Comment (ColeBollig) [inline bindings/python/htcondor2/_collector.py:43]: Should this include a `FIXME` for future documentation?

Comment (Todd-L-Miller) [reply bindings/python/htcondor2/_collector.py:43]: Yes, and I'd actually completely forgotten I hadn't finished the documentation when I asked you to review this PR.  I can do that on this PR if you want more to review, or write up a follow-up ticket.

File: src/python-bindings/classad2/classad.cpp [hunk 2]
```diff
@@ -765,3 +773,141 @@ _classad_unquote( PyObject *, PyObject * args ) {
...
+
+    auto * classAd = (ClassAd *)self->t;
+    auto * expression = (ExprTree *)expr->t;
+
+    const bool full_names = true;
+    classad::References references;
+    if(! classAd->GetInternalReferences(expression, references, full_names)) {
+        // This was a ClassAdValueError in version 1.
+        PyErr_SetString(PyExc_ValueError, "Unable to determine external references.");
+        return NULL;
```
Comment (ColeBollig) [inline src/python-bindings/classad2/classad.cpp:865]:
```suggestion
    if(! classAd->GetInternalReferences(expression, references, full_names)) {
        // This was a ClassAdValueError in version 1.
        PyErr_SetString(PyExc_ValueError, "Unable to determine internal references.");
        return NULL;
```

File: src/python-bindings/classad2/classad.cpp [hunk 3]
```diff
@@ -765,3 +773,141 @@ _classad_unquote( PyObject *, PyObject * args ) {
...
+    classad::ClassAdJsonUnParser json_up;
+    json_up.Unparse( str, (ClassAd *)handle->t );
+
+    return PyUnicode_FromString(str.c_str());
+}
+
+
+static PyObject *
+_classad_print_old( PyObject *, PyObject * args ) {
+    // _classad_print_json( self._handle )
```
Comment (ColeBollig) [inline src/python-bindings/classad2/classad.cpp:899]: Copy pasta comment mistake here... should be _classad_print_old

Comment (ColeBollig) [reply src/python-bindings/classad2/classad.cpp:899]: Just wanted to note this for whenever other changes are added

File: src/python-bindings/htcondor2/credd.cpp [hunk 4]
```diff
@@ -160,16 +160,70 @@ _credd_do_check_oauth_creds(PyObject *, PyObject * args) {
...
+    #define KEEP_PRIVS false
+
+    MyPopenTimer child;
+    if( child.start_program( producerArgs, NO_STDERR, NULL, KEEP_PRIVS ) < 0 ) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "could not run credential producer" );
+        return NULL;
+    }
+
+    #define CAREFUL_RESEARCH 20
```
Comment (ColeBollig) [inline src/python-bindings/htcondor2/credd.cpp:201]: Love this

File: src/python-bindings/classad2/classad.cpp (hunk 3 above)
Comment (Todd-L-Miller) [reply src/python-bindings/classad2/classad.cpp:899]: Right-o.

File: bindings/python/htcondor2/_collector.py (hunk 1 above)
Comment (ColeBollig) [reply bindings/python/htcondor2/_collector.py:43]: That is up to you, I am fine either way if you want to finish the documentation in this PR or a separate one.

Comment (Todd-L-Miller) [reply bindings/python/htcondor2/_collector.py:43]: This is only urgent enough to keep things more-or-less fresh in my mind, so let's set a good example and do the documentation in the first PR. ;)

File: bindings/python/htcondor2/_job_event.py [hunk 5]
```diff
@@ -9,24 +9,49 @@
...
+    undocumented.
+
+    Because all events have the ``type``, ``cluster``, ``proc``,
+    and ``timestamp`` properties, these are available as attributes;
+    see below.
+    """
+
+    # It would be nice if this signature didn't appear in the docs,
+    # since instantiating these objects shouldn't be in the API.
+    # ... there's probably a standard Pythonic way ot indicating that.
```
Comment (ColeBollig) [inline bindings/python/htcondor2/_job_event.py:27]: Just for future reference/testing `:exclude-members: __init__` may do the job. Not fully sure.

Comment (Todd-L-Miller) [reply bindings/python/htcondor2/_job_event.py:27]: Tried it and couldn't get it to work.  There's some code changes we could make to make it harder to try to call the constructor directly, but hopefully this just won't ever become an issue.

**Review by ColeBollig — APPROVED** [review]
LGTM

**Pull request merged**